compiler/src/main/scala/com/ing/baker/compiler/RecipeCompiler.scala

```diff
@@ -280,24 +285,47 @@ object RecipeCompiler {
       ++ internalEventArcs
       ++ multipleOutputFacilitatorArcs)
 
-    val petriNet: PetriNet[Place[_], Transition] = PetriNet(Graph(arcs: _*))
+    val petriNet: PetriNet[Place[_], Transition] = new PetriNet(Graph(arcs: _*))
 
     val initialMarking: Marking[Place] = petriNet.places.collect {
       case p@Place(_, FiringLimiterPlace(n)) => p -> Map((null, n))
     }.toMarking
 
+    val errors = preconditionORErrors ++ preconditionANDErrors ++ precompileErrors
+
+    val recipeId = this.generateMD5((recipe.name, petriNet, initialMarking, errors, recipe.eventReceivePeriod, recipe.retentionPeriod))
+
     val compiledRecipe = CompiledRecipe(
       name = recipe.name,
+      recipeId = recipeId,
       petriNet = petriNet,
       initialMarking = initialMarking,
-      validationErrors = preconditionORErrors ++ preconditionANDErrors ++ precompileErrors,
+      validationErrors = errors,
       eventReceivePeriod = recipe.eventReceivePeriod,
       retentionPeriod = recipe.retentionPeriod
     )
 
     postCompileValidations(compiledRecipe, validationSettings)
   }
 
+  private def generateMD5(obj: Object): String = {
```

Comment:
It looks like `obj` parameter is not used at all in this function?

I would suggest waiting for a merge of https://github.com/ing-bank/baker/pull/147

We can then implement a simple hash function without resorting to serializing with a ByteArrayOutputStream

Reply:
Agreed, my stupid mistake. Forgot to write object to the stream.

By the way, according to @Tim-Linschoten object serializer in scala automatically picks up protobuf files but it looks like it is not (just tested it). So our thoughts were to use exactly this as CompiledRecipe should be serializable with protobuf. Is it correct?

And can you give a bit more highlight on merge request #147 ? How does it help us?

Reply:
My guess is that the only reason there is no stable hashCode for 

``(recipe.name, petriNet, initialMarking, errors, recipe.eventReceivePeriod, recipe.retentionPeriod).hashCode``

is because of the `Class` type down the tree

petriNet->transition-> .. -> eventDescriptor->PrimitiveType(class)

This is resolved(removed) in #147 

Besides that, I don't think it is a good idea to serialize the entire recipe just to get a hashcode.